README.md
# 👻 ppppost

Post to Mastodon and Twitter simultaneously from the command line.

## Installation

```
$ npm install -g ashur/ppppost
```

### Requirements

- [Node 10.0](https://nodejs.org/en/download/) (or newer) & NPM

## Adding a Bot

```
$ ppppost add @tinyskylines
```

By default, bot definitions (including access tokens) are stored in a plaintext JSON file located at `~/.config/ppppost/config.json`.

> 💡 You can override this location by setting the environment variable `PPPPOST_CONFIG`

### Mastodon

You'll be prompted for the following details when configuring a bot for Mastodon:

- Instance (ex., `mastodon.social`)
- Access Token
- Visibility (`direct`, `private`, `unlisted`, or `public`)

You can generate a new Mastodon application at `https://<instance>/settings/application`, then find the `Access Token` on the application's settings page.

### Twitter

When configuring a bot for Twitter, you'll be prompted for the following:

- Consumer Token
- Consumer Secret
- Access Token
- Access Token Secret

These are available for existing Twitter applications at [https://apps.twitter.com](https://apps.twitter.com).


## Posting

To post a message, use the `to <bot>` command:

```
$ ppppost to @namogenbo --message "Happy “National Canteen Raising Month”! #NaCaRaMo 🎉"
```

> ⚠️ **Note** — ppppost does not enforce any message length limit on either Mastodon or Twitter

The `--images` option accepts a comma-separated list of file paths:

```
$ ppppost to @schneiderlens --images tail.png,teeth.png,hooves.png
```

To add captions to your images, the `--captions` option accepts a comma-separated list that corresponds to the `--images` list:

```
$ ppppost to @schneiderlens --images tail.png,teeth.png --captions "The bushy tail.","Big teeth."
```

To preserve commas in a caption, escape them with `\`:

```
$ ppppost to @schneiderlens --images tail.png,teeth.png --captions "The bushy\, bushy tail.","Big teeth."
```

By default, `ppppost` will post to all services defined for the bot. To post to just one service, use the `[-M|--mastodon]` or `[-T|--twitter]` options.


## TODO

A more fully baked version of this tool should include:

- Support for OAuth-based authentication (instead of prompting for tokens and secrets)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1] - 2019-08-16
### Added
- Support for preserving commas in captions by escaping with `\`


src/app.js
const Bot = require( './bot' );
const Config = require( './config' );
const jsonfile = require( 'jsonfile' );
const path = require( 'path' );

class App
{
	constructor()
	{
		this._bots = [];
	}

	/**
	 * Add a bot to the list of bots
	 * @param {Bot} bot
	 */
	addBot( bot )
	{
		if( bot instanceof Bot )
		{
			this._bots.push( bot );
		}
		else
		{
			throw new Error( 'Invalid argument type: Bot required' );
		}
	}

	/**
	 * Find a single bot
	 * @param {string} name - Name of bot
	 * @param {Bot}
	 */
	bot( name )
	{
		return this._bots.find( bot => bot.name == name );
	}

	/**
	 * Return an array of Bot objects, sorted by name
	 * @return {Bot[]} bots - An array of Bot objects
	 */
	get bots()
	{
		return this._bots.sort( (a, b) =>
		{
			if( a.name < b.name )
			{
				return -1;
			}

			return 1;
		});
	}

	/**
	 * Return whether the list of bots contains a specific bot name
	 * @param {string} name
	 * @return {boolean}
	 */
	hasBot( name )
	{
		return this._bots.find( bot => bot.name == name ) != null;
	}

	/**
	 * Remove a bot from the list of bots
	 * @param {string} name
	 */
	removeBot( name )
	{
		let index = this._bots.findIndex( bot => bot.name == name );
		this._bots.splice( index, 1 );
	}

	/**
	 * Save the bot list to a JSON file
	 * @param {string} configPath
	 * @return {Promise}
	 */
	save( configPath )
	{
		let config = new Config( { version: "1.0", bots: this._bots } );
		return config.write( configPath );
	}

	/**
	 * Replaces an existing bot with a new instance
	 * @param {string} name
	 * @param {Bot} bot
	 */
	updateBot( name, bot )
	{
		let index = this._bots.findIndex( bot => bot.name == name );
		this._bots.splice( index, 1, bot );
	}
}

module.exports = App;


src/bot.js
const Mastodon = require( './mastodon' );
const Twitter = require( './twitter' );

class Bot
{
	constructor( { name, mastodon, twitter } )
	{
		this.name = name;

		if( mastodon )
		{
			this.mastodon = new Mastodon( mastodon );
		}
		if( twitter )
		{
			this.twitter = new Twitter( twitter );
		}
	}

	/**
	 * Return whether Mastodon credentials are defined
	 * @return {boolean}
	 */
	hasMastodon()
	{
		return this.mastodon != null;
	}

	/**
	 * Return whether Twitter credentials are defined
	 * @return {boolean}
	 */
	hasTwitter()
	{
		return this.twitter != null;
	}

	/**
	 * Attempt to post a new status to Mastodon. Resolves Promise with ID of
	 * the newly created toot if successful.
	 *
	 * @param {Object} status - The status to be posted
	 * @param {string} status.message - The status message [optional]
	 * @param {array} status.media - Array of local paths to images [optional]
	 * @param {string} status.visibility - Override default bot visibility
	 * 
	 * @return {Promise}
	 */
	toot( status )
	{
		return new Promise( (resolve, reject) =>
		{
			if( !this.hasMastodon() )
			{
				reject( `'${this.name}' has not been configured for Mastodon` );
			}

			this.mastodon.post( status )
				.then( response =>
				{
					resolve( response.data.id );
				})
				.catch( error =>
				{
					reject( error );
				});
		})
	}

	/**
	 * Attempt to post a new status to Twitter. Resolves Promise with ID of
	 * the newly created tweet if successful.
	 *
	 * @param {Object} status - The status to be posted
	 * @param {string} status.message - The status message [optional]
	 * @param {array} status.media - Array of local paths to images [optional]
	 * 
	 * @return {Promise}
	 */
	tweet( status )
	{
		return new Promise( (resolve, reject) =>
		{
			if( !this.hasTwitter() )
			{
				reject( `'${this.name}' has not been configured for Twitter` );
			}

			this.twitter.post( status )
				.then( response =>
				{
					resolve( response.data.id );
				})
				.catch( error =>
				{
					reject( error );
				});
		});
	}
}

module.exports = Bot;


src/config.js
const jsonfile = require( 'jsonfile' );
const path = require( 'path' );
const mkdirp = require( 'mkdirp' );

class Config
{
	constructor( { version="1.0", bots=[] } = {} )
	{
		this.version = version;

		if( version == "1.0" )
		{
			this.bots = bots;
		}
	}

	/**
	 * Read a JSON config file and create a new Config instance
	 * @param {string} configPath
	 * @return {boolean}
	 */
	static read( configPath )
	{
		return new Promise( (resolve, reject) =>
		{
			let absolutePath = path.resolve( configPath.replace( '~', process.env.HOME ) );

			jsonfile.readFile( absolutePath )
				.then( data =>
				{
					data.path = absolutePath;
					resolve( new Config( data ) );
				})
				.catch( error =>
				{
					reject( error.message );
				});
		});
	}

	/**
	 * Write the current instance's properties to a JSON config file
	 * @param {string} configPath
	 * @return {Promise}
	 */
	write( configPath )
	{
		return new Promise( (resolve, reject) =>
		{
			let data = {};

			if( this.version == "1.0" )
			{
				data.version = this.version;
				data.bots = this.bots;
			}

			let absolutePath = path.resolve( configPath.replace( '~', process.env.HOME ) );

			// Attempt to create the parent folder structure in case it doesn't
			// already exist.

			mkdirp( path.dirname( absolutePath ), { recursive: true }, error =>
			{
				if( !error )
				{
					jsonfile.writeFile( absolutePath, data )
						.then( () =>
						{
							resolve();
						})
						.catch( error =>
						{
							reject( error );
						});
				}
				else
				{
					reject( error );
				}
			});
		});
	}
}

Config.DEFAULT_PATH = '~/.config/ppppost/config.json';

module.exports = Config;


src/twitter.js
const fs = require( 'fs' );
const T = require( 'twit' );

class Twitter
{
	constructor( { consumer_key, consumer_secret, access_token, access_token_secret }  )
	{
		this.consumer_key = consumer_key;
		this.consumer_secret = consumer_secret;
		this.access_token = access_token;
		this.access_token_secret = access_token_secret;
	}

	/**
	 * Post a new status to Twitter
	 * @param {Object} object
	 * @param {string} [object.message] - A status message string
	 * @param {string[]} [object.captions] - An array of captions corresponding to media
	 * @param {string[]} [object.media] - An array of path strings
	 * @return {Promise}
	 */
	post( { message='', media=[], captions=[] } = {} )
	{
		if( message == '' && media.length == 0 )
		{
			throw new Error( 'You must provide either a status message or media, or both.' );
		}

		let t = new T({
			consumer_key: this.consumer_key,
			consumer_secret: this.consumer_secret,
			access_token: this.access_token,
			access_token_secret: this.access_token_secret,
			timeout_ms: 60*1000,
		});

		let mediaIds = [];
		let mediaUploads = media.map( (path, index) =>
		{
			let media = {
				media_data: fs.readFileSync( path , { encoding: 'base64' } )
			};

			return t.post( 'media/upload', media )
				.then( response =>
				{
					let mediaId = response.data.media_id_string
					mediaIds.push( mediaId );

					if( captions[index] )
					{
						let metadata = {
							media_id: mediaId,
							alt_text: {
								text: captions[index]
							}
						};

						return t.post( 'media/metadata/create', metadata );
					}
				})
		});

		return Promise.all( mediaUploads )
			.then( () =>
			{
				let tweet = {};

				if( message != '' )
				{
					tweet.status = message;
				}

				if( mediaIds.length > 0 )
				{
					tweet.media_ids = mediaIds;
				}

				return t.post( 'statuses/update', tweet );
			});
	}
}

module.exports = Twitter;


src/helpers/prompt.js
/**
 * @param {Object} object
 * @param {readline.Interface} [object.rl]
 * @param {string} [object.message]
 * @param {boolean} [object.required] - Re-prompt if empty answer is supplied
 * @param {boolean} [object.defaultAnswer]
 * @param {number} [object.defaultAnswerMaxLength] - Number of characters to truncate to
 * @return {Promise}
 */
function prompt( {rl, message, required=false, defaultAnswer, defaultAnswerMaxLength} )
{
	return new Promise( (resolve, reject) =>
	{
		if( !rl )
		{
			reject( "Readline interface property 'rl' required." );
		}

		let defaultAnswerString;
		if( defaultAnswerMaxLength )
		{
			defaultAnswerString = truncate( defaultAnswer, defaultAnswerMaxLength );
		}
		else
		{
			defaultAnswerString = defaultAnswer;
		}
		defaultAnswerString = defaultAnswer ? ` [${defaultAnswerString}]` : '';

		rl.question( `${message}${defaultAnswerString}: `, answer =>
		{
			if( answer == '' && required )
			{
				resolve( prompt({
					rl: rl,
					message: message,
					required: required,
					defaultAnswer: defaultAnswer,
					defaultAnswerMaxLength: defaultAnswerMaxLength,
				}));
			}
			if( answer == '' && defaultAnswer )
			{
				answer = defaultAnswer;
			}

			resolve( answer );
		});
	});
}

/** 
 * @param {string} string
 * @param {number} length
 * @return {string}
 */
function truncate( string, length )
{
	if( string == null )
	{
		return null;
	}

	if( string.length <= length )
	{
		return string;
	}

	let ellipsisLength = "...".length;
	let leftLength = Math.ceil( (length - ellipsisLength) / 2 );
	let rightLength = Math.floor( (length - ellipsisLength) / 2 );
	
	let truncated = `${string.substring(0, leftLength)}...${string.substring(string.length - rightLength)}`;

	return truncated;
}

module.exports = prompt;


src/helpers/app-from-config-path.js
const App = require( '../app' );
const Bot = require( '../bot' );
const Config = require( '../config' );

let app = new App();

module.exports.bootstrap = configPath =>
{
	return new Promise( (resolve, reject) =>
	{
		Config.read( configPath )
			.then( config =>
			{
				config.bots.forEach( data =>
				{
					app.addBot( new Bot( data ) );
				});

				resolve( app );
			})
			.catch( error =>
			{
				// The file doesn't exist, so let's try to create it

				let config = new Config();
				config.write( configPath )
					.then( () =>
					{
						config.bots.forEach( data =>
						{
							app.addBot( new Bot( data ) );
						});

						resolve( app );
					})
					.catch( error =>
					{
						reject( `Could not read or write to ${configPath}: ${error.message}` );
					})
			});
	});
};

module.exports.Config = Config;


src/mastodon.js
const fs = require( 'fs' );
const M = require( 'mastodon' );

class Mastodon
{
	constructor( { api_url, access_token, visibility='unlisted' }  )
	{
		this.access_token = access_token;
		this.api_url = api_url;
		this.visibility = visibility;
	}

	/**
	 * Post a new status to Mastodon
	 * @param {Object} object
	 * @param {string} [object.message] - A status message string
	 * @param {string[]} [object.media] - An array of path strings
	 * @param {string[]} [object.captions] - An array of captions corresponding to media
	 * @param {string} [object.visibility] - Override default visibility
	 * @return {Promise}
	 */
	post( { message='', media=[], captions=[], visibility=this.visibility } = {} )
	{
		if( message == '' && media.length == 0 )
		{
			throw new Error( 'You must provide either a status message or media, or both.' );
		}

		let m = new M({
			access_token: this.access_token,
			api_url: this.api_url,
			timeout_ms: 60*1000,
		});

		let mediaIds = [];
		let mediaUploads = media.map( (path, index) =>
		{
			let media = {
				file: fs.createReadStream( path )
			};

			if( captions[index] )
			{
				media.description = captions[index];
			}

			return m.post( 'media', media )
				.then( response =>
				{
					mediaIds.push( response.data.id );
				});
		});

		return Promise.all( mediaUploads )
			.then( () =>
			{
				// We need to send along 'status' even if it's an empty string,
				// or Mastodon gets grumpy.

				let toot = {
					status: message,
					visibility: visibility,
				};

				if( mediaIds.length > 0 )
				{
					toot.media_ids = mediaIds;
				}

				return m.post( 'statuses', toot );
			});
	}
}

module.exports = Mastodon;
